core/templates/components/button-directives/create-activity-button.component.spec.ts

```diff
@@ -0,0 +1,289 @@
+    component.initCreationProcess();
+    expect(component.creationInProgress).toBe(true);
+    expect(explorationCreationServiceSpy).toHaveBeenCalledTimes(1);
+
+    component.initCreationProcess();
+    expect(explorationCreationServiceSpy).toHaveBeenCalledTimes(1);
+  });
+
+  it('should create new exploration if user cannot create collections', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = false;
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    component.initCreationProcess();
+
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should redirect user to a new exploration when user clicks' +
+    ' create button and is not on creator dashboard page', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = true;
+    spyOn(urlService, 'getPathname').and.returnValue(
+      'not/creator-dashboard');
+    spyOnProperty(windowRef, 'nativeWindow').and.returnValue({
+      location: {
+        replace: (val: string) => {}
+      }
+    });
+    const replaceSpy = spyOn(windowRef.nativeWindow.location, 'replace');
+
+    component.initCreationProcess();
+
+    expect(replaceSpy).toHaveBeenCalledWith('/creator-dashboard?mode=create');
+  });
+
+  it('should not redirect user but open a create activity modal if user' +
+    ' can create collections and is on creator dashboard page', () => {
+    component.creationInProgress = false;
+    component.canCreateCollections = true;
+    spyOn(urlService, 'getPathname').and.returnValue(
+      '/creator-dashboard');
+    const modalSpy = spyOn(ngbModal, 'open').and.callFake((dlg, opt) => {
+      return <NgbModalRef>({
+        result: Promise.resolve('success')
+      });
+    });
+    const replaceSpy = spyOn(windowRef.nativeWindow.location, 'replace');
+
+    component.initCreationProcess();
+
+    expect(replaceSpy).not.toHaveBeenCalled();
+    expect(modalSpy).toHaveBeenCalled();
+  });
+
+  it('should handle modal\'s failure callback' +
+    ' and stop creation process', fakeAsync(() => {
+    component.creationInProgress = false;
+    component.canCreateCollections = true;
+    spyOn(urlService, 'getPathname').and.returnValue(
+      '/creator-dashboard');
+    const modalSpy = spyOn(ngbModal, 'open').and.callFake((dlg, opt) => {
+      return <NgbModalRef>({
+        result: Promise.reject('cancel')
+      });
+    });
+
+    component.initCreationProcess();
+    tick();
+
+    expect(modalSpy).toHaveBeenCalled();
+    expect(component.creationInProgress).toBe(false);
+  }));
+
+  it('should show upload exploration modal', () => {
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'showUploadExplorationModal');
+
+    component.showUploadExplorationModal();
+
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should redirect to login page when user creates exploration' +
```

Comment:
You'd probably want to add another test after this to verify that the method on siteAnalyticsService is called. That's an example of an external side effect that should be validated.

Reply:
Done

Reply:
No, I explicitly asked for a separate test here because you are testing a separate thing. This test checks redirection. A separate test should check analytics.

Again, have the test body match the test description. This is a very important fundamental point.

Reply:
Ah, alright. Added another test